### PR TITLE
GCP setup

### PR DESCRIPTION
**Pull request by jiya-zhang**

### CODE REVIEW

File: axlearn/cloud/gcp/project_setup.sh [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# This is a script to set up a brand new GCP project before you use AXLearn tools"""
```
Comment (markblee) [inline axlearn/cloud/gcp/project_setup.sh:1]:
```suggestion
# Sets up a brand new GCP project for AXLearn. See also the "Getting Started" docs linked in the main readme.
```
Should we move it under https://github.com/apple/axlearn/tree/main/axlearn/cloud/gcp/scripts?

File: axlearn/cloud/gcp/project_setup.sh [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+# Usage: 
+#    # fill out environment variables below
+#    chmod +x project_setup.sh
+#    ./project_setup.sh
+
+
+#!/bin/sh
+
+set -e
+#set -x
```
Comment (markblee) [inline axlearn/cloud/gcp/project_setup.sh:12]: Is this commented line intentional?

File: docs/01-start.md [hunk 3]
```diff
@@ -653,7 +653,7 @@ AXLearn comes with tooling for provisioning and launching training on public clo
 ### Pre-requisites
 
 We assume you have:
-1. A Google Cloud Platform (GCP) project with TPU quota.
+1. A Google Cloud Platform (GCP) project with TPU quota. To set up a brand new GCP project, please follow [these instructions](05-gcp-setup.md).
```
Comment (markblee) [inline docs/01-start.md:656]: Maybe we can directly link to the script here. The script itself should have instructions on how it should be run at the top of the file, mitigating possible out-of-sync issues. WDYT?

**Review by markblee — COMMENTED** [review]
Thanks @jiya-zhang !

File: axlearn/cloud/gcp/project_setup.sh (hunk 2 above)
Comment (jiya-zhang) [reply axlearn/cloud/gcp/project_setup.sh:12]: I figured you can test both and see whether you want to include this line. From my own experience, including "set -x" made my terminal output confusing...

File: docs/01-start.md (hunk 3 above)
Comment (jiya-zhang) [reply docs/01-start.md:656]: Makes sense, then we wouldn't need the "05-gcp-setup.md" file

File: axlearn/cloud/gcp/project_setup.sh (hunk 1 above)
Comment (jiya-zhang) [reply axlearn/cloud/gcp/project_setup.sh]: Yes... moved it in new commit. (But I forgot to commit suggestion first, so you might have to do it again...)

File: docs/01-start.md (hunk 3 above)
Comment (markblee) [reply docs/01-start.md:656]: Yeah, seems like we can point users directly to the script.

File: axlearn/cloud/gcp/project_setup.sh (hunk 2 above)
Comment (markblee) [reply axlearn/cloud/gcp/project_setup.sh]: If it's not adding much benefits we can remove it.

Comment (jiya-zhang) [reply axlearn/cloud/gcp/project_setup.sh]: cool, edited

File: docs/01-start.md (hunk 3 above)
Comment (jiya-zhang) [reply docs/01-start.md]: made edits in the latest commit

**Review by markblee — APPROVED** [review]

File: axlearn/cloud/gcp/scripts/project_setup.sh [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+gcloud alpha billing projects link $PROJECT_ID --billing-account=$BILLING_ACCOUNT_ID
+
+# Step 3: Enable services
+gcloud services enable compute.googleapis.com
+gcloud services enable tpu.googleapis.com
+
+# Step 4: Create an auto mode VPC
+# https://cloud.google.com/vpc/docs/create-modify-vpc-networks#create-auto-network 
+gcloud compute networks create $NETWORK_NAME --subnet-mode=auto
+# Note: certain TPUs (eg. v4) are located in a private region (eg. us-central2).
```
Comment (ryanoceros-g) [inline axlearn/cloud/gcp/scripts/project_setup.sh:48]: i'd recommend putting notes on TPU quota/region at or near the top

File: axlearn/cloud/gcp/scripts/project_setup.sh [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+export SERVICE_ACCOUNT_ID=${SERVICE_ACCOUNT_NAME}@${PROJECT_ID}.iam.gserviceaccount.com 
+
+# Step 1: Create a GCP project
+# https://cloud.google.com/resource-manager/docs/creating-managing-projects
+gcloud projects create $PROJECT_ID --organization=$ORGANIZATION_ID
+gcloud config set project $PROJECT_ID
+
+# Step 2: Link the project to a billing account
+# https://cloud.google.com/billing/docs/how-to/modify-project
+gcloud alpha billing projects link $PROJECT_ID --billing-account=$BILLING_ACCOUNT_ID
```
Comment (ryanoceros-g) [inline axlearn/cloud/gcp/scripts/project_setup.sh:39]: Is there some reason we are using alpha here?

**Review by ryanoceros-g — CHANGES_REQUESTED** [review]
Left several comments in this review, but primarily hoping to have gcloud alpha command changed before merge if possible.

File: axlearn/cloud/gcp/scripts/project_setup.sh (hunk 5 above)
Comment (jiya-zhang) [reply axlearn/cloud/gcp/scripts/project_setup.sh]: I checked documentation and changed this to non-alpha command

File: docs/01-start.md [hunk 6]
```diff
@@ -653,7 +653,7 @@ AXLearn comes with tooling for provisioning and launching training on public clo
 ### Pre-requisites
 
 We assume you have:
-1. A Google Cloud Platform (GCP) project with TPU quota.
+1. A Google Cloud Platform (GCP) project with TPU quota. To set up a brand new GCP project, please run [this script](../axlearn/cloud/gcp/scripts/project_setup.sh).
```
[the remark this answers is not in the document]
Comment (jiya-zhang) [reply docs/01-start.md]: Adjusted the order of the prerequisites to make it clearer

**Review by ryanoceros-g — APPROVED** [review]
LGTM

File: axlearn/cloud/gcp/scripts/project_setup.sh [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+export SERVICE_ACCOUNT_ID=${SERVICE_ACCOUNT_NAME}@${PROJECT_ID}.iam.gserviceaccount.com 
+
+# Step 1: Create a GCP project
+# https://cloud.google.com/resource-manager/docs/creating-managing-projects
+gcloud projects create $PROJECT_ID --organization=$ORGANIZATION_ID
+gcloud config set project $PROJECT_ID
+
+# Step 2: Link the project to a billing account
+# https://cloud.google.com/billing/docs/how-to/modify-project
+gcloud billing projects link $PROJECT_ID --billing-account=$BILLING_ACCOUNT_ID
```
Comment (samos123) [inline axlearn/cloud/gcp/scripts/project_setup.sh:42]: I think we should assume the project has been created and a billing account has been linked. In most organizations this is a different team.

File: axlearn/cloud/gcp/scripts/project_setup.sh [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+# Step 4: Create an auto mode VPC
+# https://cloud.google.com/vpc/docs/create-modify-vpc-networks#create-auto-network 
+gcloud compute networks create $NETWORK_NAME --subnet-mode=auto
+# Note: certain TPUs (eg. v4) are located in a private region (eg. us-central2). 
+# You will need to request quota before creating a subnet in that case:
+# https://cloud.google.com/tpu/docs/setup-gcp-account#prepare-to-request 
+
+# Step 5: Create GCS buckets
+# https://cloud.google.com/storage/docs/creating-buckets#storage-create-bucket-cli
+gcloud storage buckets create gs://$PERMANENT_BUCKET_NAME --location=$BUCKET_REGION --uniform-bucket-level-access
```
Comment (samos123) [inline axlearn/cloud/gcp/scripts/project_setup.sh:57]: bundle this in an if statement that checks if the resource already exists, so it becomes idempotent

File: axlearn/cloud/gcp/scripts/project_setup.sh [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+gcloud storage buckets create gs://$PRIVATE_BUCKET_NAME --location=$BUCKET_REGION --uniform-bucket-level-access
```
Comment (samos123) [inline axlearn/cloud/gcp/scripts/project_setup.sh:58]: bundle this in an if statement that checks if the resource already exists, so it becomes idempotent

File: axlearn/cloud/gcp/scripts/project_setup.sh [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+# https://cloud.google.com/billing/docs/how-to/modify-project
+gcloud billing projects link $PROJECT_ID --billing-account=$BILLING_ACCOUNT_ID
+
+# Step 3: Enable services
+gcloud services enable compute.googleapis.com
+gcloud services enable tpu.googleapis.com
+
+# Step 4: Create an auto mode VPC
+# https://cloud.google.com/vpc/docs/create-modify-vpc-networks#create-auto-network 
+gcloud compute networks create $NETWORK_NAME --subnet-mode=auto
```
Comment (samos123) [inline axlearn/cloud/gcp/scripts/project_setup.sh:50]: bundle this in an if statement that checks if the resource already exists, so it becomes idempotent

**Review by samos123 — CHANGES_REQUESTED** [review]
All create statements should have an if statement to check if the resource already exists, that way you can rerun the script if needed when one of the resources failed to create. It also makes it possible to run this script when some of the resources already exist.

File: axlearn/cloud/gcp/scripts/project_setup.sh [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+# Step 4: Create an auto mode VPC
+# https://cloud.google.com/vpc/docs/create-modify-vpc-networks#create-auto-network
+gcloud compute networks create $NETWORK_NAME --subnet-mode=auto
+# Note: certain TPUs (eg. v4) are located in a private region (eg. us-central2).
+# You will need to request quota before creating a subnet in that case:
+# https://cloud.google.com/tpu/docs/setup-gcp-account#prepare-to-request
+
+# Step 5: Create GCS buckets
+# https://cloud.google.com/storage/docs/creating-buckets#storage-create-bucket-cli
+gcloud storage buckets create gs://$PERMANENT_BUCKET_NAME --location=$BUCKET_REGION --uniform-bucket-level-access
```
Comment (samos123) [inline axlearn/cloud/gcp/scripts/project_setup.sh:57]: This is how you would do this:
```
if ! gcloud storage buckets describe "${PERMANENT_BUCKET_NAME}" -q >/dev/null; then
  gcloud storage buckets create gs://$PERMANENT_BUCKET_NAME --location=$BUCKET_REGION --uniform-bucket-level-access
fi
```

**Review by samos123 — APPROVED** [review]

**Review by markblee — APPROVED** [review]

**Pull request merged**